### ISSUE TITLE
Workflows/activities to post event notifications that could be listened to

### ISSUE THREAD

**Issue by sstro** [temporalio/temporal#3709]
We would like to listen to Temporal workflow/activity lifecycle events (e.g. _WorkflowExecutionStarted_, _ActivityTaskStarted_, _ActivityTaskCompleted_), apply some filtering and transformation and then persist the events to our own database. Business users could query this database directly instead of querying Temporal and transforming/aggregating information at the runtime.

An additional requirement would be ability to distinguish workflow / activity replay / retry events from "first execution" events.

It would be nice if it were possible to attach some kind of listener to a workflow / an activity at the worker level.

We've considered workflow/activity inbound/outbound call interceptors, however, they only seem to fire on a subset of events (start/finish). Additionally, workflow interceptors do not distinguish first start and replay events.

We would be interested in having the functionality available in Java SDK.

Related discussions in the support forum: 
- https://community.temporal.io/t/how-to-watch-all-workflows-events-inside-a-namespace/4658
- https://community.temporal.io/t/way-to-listen-on-workflow-lifecycle-events/5887
- https://community.temporal.io/t/consume-temporal-io-events/1630
- https://community.temporal.io/t/can-i-attach-a-listener-to-workflow-execution/4275

**Comment by mindaugasrukas** [comment]
Is this a use case for having the ability to create a custom UI (user interface)?

**Comment by sstro** [comment]
Not exactly, it's for integration with our own process information system.

**Comment by yiminc** [comment]
We have 2 options currently available:
1) use interceptor to intercept workflow/activity, example  https://github.com/temporalio/samples-go/tree/main/interceptor
2) intercept calls using a proxy, example: https://github.com/temporalio/samples-go/tree/main/grpc-proxy

**Comment by sstro** [comment]
We've considered workflow/activity inbound/outbound call interceptors, however, they only seem to fire on a subset of events (start/finish). Additionally, workflow interceptors do not distinguish first start and replay events, which is important for our use case.

Is the proxy supported in Java SDK ?

**Comment by Quinn-With-Two-Ns** [comment]
Proxy and interceptors are not a complete solution for this issue since they are not invoked on server side workflow completions like timeout or termination.
